File: src/main/java/magma/list/JdkList.java
package magma.list;

import java.util.ArrayList;
import java.util.List;

/** Default ListLike backed by java.util.ArrayList. */
public class JdkList<T> implements ListLike<T> {
    private final List<T> list;

    private JdkList(List<T> list) {
        this.list = list;
    }

    /** Create an empty list. */
    public static <T> JdkList<T> create() {
        return new JdkList<>(new ArrayList<>());
    }

    /** Wrap an existing java.util.List. */
    public static <T> JdkList<T> wrap(List<T> list) {
        return new JdkList<>(list);
    }

    @Override
    public void add(T value) {
        list.add(value);
    }

    @Override
    public T get(int index) {
        return list.get(index);
    }

    @Override
    public void set(int index, T value) {
        list.set(index, value);
    }

    @Override
    public int size() {
        return list.size();
    }

    @Override
    public ListIter<T> iterator() {
        java.util.Iterator<T> it = list.iterator();
        return new ListIter<T>() {
            @Override
            public boolean hasNext() {
                return it.hasNext();
            }

            @Override
            public T next() {
                return it.next();
            }
        };
    }
}

File: src/main/node/magma/list/JdkList.ts
import ArrayList from "../../java/util/ArrayList";
import List from "../../java/util/List";
/** Default ListLike backed by java.util.ArrayList. */
export default class JdkList<T> implements ListLike<T> {
    private readonly list: List<T>;

    JdkList(list: List<T>): private {
        // TODO
    }

    /** Create an empty list. */
    create(): JdkList<T> {
        return new JdkList<T>(new ArrayList<T>());
    }

    /** Wrap an existing java.util.List. */
    wrap(list: List<T>): JdkList<T> {
        return new JdkList<T>(list);
    }

    @Override
    add(value: T): void {
        list.add(value);
    }

    @Override
    get(index: number): T {
        return list.get(index);
    }

    @Override
    set(index: number, value: T): void {
        list.set(index, value);
    }

    @Override
    size(): number {
        return list.size();
    }

    @Override
    iterator(): ListIter<T> {
        let it : java.util.Iterator<T> = list.iterator();
        return new ListIter<T>();
        // TODO
        public boolean hasNext();
        return it.hasNext();
        // TODO
        // TODO
        public T next();
        return it.next();
        // TODO
        // TODO
    }
}
